File: exercises/cpp/04b_left_recursion/left_recursion.cpp
/*
================================================================================
Exercise 4b: Left Recursion Elimination (C++ Implementation)
================================================================================

CONCEPT:
Left recursion in a context-free grammar occurs when a non-terminal A can
derive a sentential form starting with itself: A =>+ A alpha. This is
problematic for top-down (recursive descent / LL) parsers because they would
enter infinite loops trying to expand A. Left recursion comes in two forms:

  - Direct left recursion:   A -> A alpha | beta
  - Indirect left recursion: A -> B alpha, B -> A beta

ALGORITHM:
For direct left recursion:
  Original:    A -> A alpha1 | A alpha2 | ... | beta1 | beta2 | ...
  Transformed: A  -> beta1 A' | beta2 A' | ...
               A' -> alpha1 A' | alpha2 A' | ... | epsilon

For indirect left recursion, we use the ordering algorithm:
  1. Order non-terminals A1, A2, ..., An.
  2. For i = 1 to n:
       For j = 1 to i-1:
         Substitute Aj in Ai productions.
       Eliminate direct left recursion from Ai.

COMPLEXITY:
  Time:  O(n^2 * p) where n = number of non-terminals, p = max productions
  Space: O(n * p) for storing the grammar

EXAMPLE:
  Input:   E -> E+T | T,  T -> T*F | F,  F -> (E) | id
  Output:  E  -> TE'
           E' -> +TE' | epsilon
           T  -> FT'
           T' -> *FT' | epsilon
           F  -> (E) | id
================================================================================
*/

#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <algorithm>

using std::string;
using std::vector;
using std::cout;
using std::endl;

/* A production is a list of symbols (strings) */
using Production = vector<string>;

struct NonTerminal {
    string name;
    vector<Production> productions;
};

/* Grammar is an ordered list of non-terminals */
using Grammar = vector<NonTerminal>;

/* ---- Parsing and printing utilities ---- */

Production parse_production(const string &text) {
    Production prod;
    std::istringstream iss(text);
    string symbol;
    while (iss >> symbol) {
        prod.push_back(symbol);
    }
    return prod;
}

string production_to_string(const Production &prod) {
    string result;
    for (size_t i = 0; i < prod.size(); i++) {
        if (i > 0) result += " ";
        result += prod[i];
    }
    return result;
}

int find_nonterminal(const Grammar &grammar, const string &name) {
    for (size_t i = 0; i < grammar.size(); i++) {
        if (grammar[i].name == name) return (int)i;
    }
    return -1;
}

Grammar parse_grammar(const string &text) {
    Grammar grammar;
    std::istringstream iss(text);
    string line;

    while (std::getline(iss, line)) {
        /* Trim whitespace */
        size_t start = line.find_first_not_of(" \t\r\n");
        if (start == string::npos) continue;
        line = line.substr(start);
        if (line.empty() || line[0] == '#') continue;

        size_t arrow_pos = line.find("->");
        if (arrow_pos == string::npos) continue;

        string lhs = line.substr(0, arrow_pos);
        /* Trim lhs */
        lhs.erase(lhs.find_last_not_of(" \t") + 1);
        lhs.erase(0, lhs.find_first_not_of(" \t"));

        string rhs = line.substr(arrow_pos + 2);

        /* Split by | */
        vector<Production> productions;
        std::istringstream rhs_stream(rhs);
        string segment;
        /* Manual split by | since getline with '|' works */
        size_t pos = 0;
        while (pos < rhs.size()) {
            size_t pipe = rhs.find('|', pos);
            if (pipe == string::npos) pipe = rhs.size();
            string alt = rhs.substr(pos, pipe - pos);
            Production prod = parse_production(alt);
            if (!prod.empty()) {
                productions.push_back(prod);
            }
            pos = pipe + 1;
        }

        int idx = find_nonterminal(grammar, lhs);
        if (idx >= 0) {
            for (auto &p : productions)
                grammar[idx].productions.push_back(p);
        } else {
            NonTerminal nt;
            nt.name = lhs;
            nt.productions = productions;
            grammar.push_back(nt);
        }
    }
    return grammar;
}

void print_grammar(const Grammar &grammar) {
    for (const auto &nt : grammar) {
        cout << "  " << nt.name << " -> ";
        for (size_t j = 0; j < nt.productions.size(); j++) {
            if (j > 0) cout << " | ";
            cout << production_to_string(nt.productions[j]);
        }
        cout << endl;
    }
}

/* ---- Left recursion elimination ---- */

/*
 * Eliminate direct left recursion for grammar[nt_idx].
 * Returns the new non-terminal (A') to insert, or empty if no recursion found.
 */
NonTerminal eliminate_direct_left_recursion(Grammar &grammar, int nt_idx) {
    NonTerminal &nt = grammar[nt_idx];
    vector<Production> recursive_parts;     /* alpha parts */
    vector<Production> non_recursive_parts; /* beta parts */

    for (const auto &prod : nt.productions) {
        if (!prod.empty() && prod[0] == nt.name) {
            /* Left-recursive: extract alpha (everything after A) */
            Production alpha(prod.begin() + 1, prod.end());
            recursive_parts.push_back(alpha);
        } else {
            non_recursive_parts.push_back(prod);
        }
    }

    NonTerminal new_nt;
    if (recursive_parts.empty()) {
        return new_nt;  /* No left recursion; new_nt.name is empty */
    }

    string new_name = nt.name + "'";
    new_nt.name = new_name;

    /* A -> beta1 A' | beta2 A' | ... */
    nt.productions.clear();
    for (auto &beta : non_recursive_parts) {
        Production new_prod;
        if (beta.size() == 1 && beta[0] == "epsilon") {
            new_prod.push_back(new_name);
        } else {
            new_prod = beta;
            new_prod.push_back(new_name);
        }
        nt.productions.push_back(new_prod);
    }

    /* A' -> alpha1 A' | alpha2 A' | ... | epsilon */
    for (auto &alpha : recursive_parts) {
        Production new_prod = alpha;
        new_prod.push_back(new_name);
        new_nt.productions.push_back(new_prod);
    }
    new_nt.productions.push_back({"epsilon"});

    return new_nt;
}

/*
 * Substitute: replace all occurrences of grammar[idx_j].name at the start
 * of productions of grammar[idx_i] with the actual productions of idx_j.
 */
void substitute_productions(Grammar &grammar, int idx_i, int idx_j) {
    const string &name_j = grammar[idx_j].name;
    vector<Production> new_prods;

    for (const auto &prod : grammar[idx_i].productions) {
        if (!prod.empty() && prod[0] == name_j) {
            Production gamma(prod.begin() + 1, prod.end());
            for (const auto &delta : grammar[idx_j].productions) {
                Production new_prod;
                if (delta.size() == 1 && delta[0] == "epsilon") {
                    if (!gamma.empty()) {
                        new_prod = gamma;
                    } else {
                        new_prod.push_back("epsilon");
                    }
                } else {
                    new_prod = delta;
                    new_prod.insert(new_prod.end(), gamma.begin(), gamma.end());
                }
                new_prods.push_back(new_prod);
            }
        } else {
            new_prods.push_back(prod);
        }
    }

    grammar[idx_i].productions = new_prods;
}

Grammar eliminate_left_recursion(Grammar grammar) {
    /* Save original non-terminal names */
    vector<string> orig_names;
    for (const auto &nt : grammar) {
        orig_names.push_back(nt.name);
    }
    int orig_count = (int)orig_names.size();

    for (int i = 0; i < orig_count; i++) {
        for (int j = 0; j < i; j++) {
            int cur_i = find_nonterminal(grammar, orig_names[i]);
            int cur_j = find_nonterminal(grammar, orig_names[j]);
            if (cur_i >= 0 && cur_j >= 0) {
                substitute_productions(grammar, cur_i, cur_j);
            }
        }

        int cur_i = find_nonterminal(grammar, orig_names[i]);
        if (cur_i >= 0) {
            NonTerminal new_nt = eliminate_direct_left_recursion(grammar, cur_i);
            if (!new_nt.name.empty()) {
                /* Insert A' right after A */
                grammar.insert(grammar.begin() + cur_i + 1, new_nt);
            }
        }
    }

    return grammar;
}

int main() {
    cout << "======================================================================" << endl;
    cout << "  Exercise 4b: Left Recursion Elimination" << endl;
    cout << "======================================================================" << endl;

    /* --- Example 1: Expression grammar --- */
    cout << "\n--- Example 1: Direct Left Recursion ---\n" << endl;
    Grammar grammar1 = parse_grammar(
        "E -> E + T | T\n"
        "T -> T * F | F\n"
        "F -> ( E ) | id\n"
    );
    cout << "Original Grammar:" << endl;
    print_grammar(grammar1);

    Grammar result1 = eliminate_left_recursion(grammar1);
    cout << "\nAfter Left Recursion Elimination:" << endl;
    print_grammar(result1);

    /* --- Example 2: Indirect left recursion --- */
    cout << "\n----------------------------------------------------------------------" << endl;
    cout << "\n--- Example 2: Indirect Left Recursion ---\n" << endl;
    Grammar grammar2 = parse_grammar(
        "S -> A a | b\n"
        "A -> S c | d\n"
    );
    cout << "Original Grammar:" << endl;
    print_grammar(grammar2);

    Grammar result2 = eliminate_left_recursion(grammar2);
    cout << "\nAfter Left Recursion Elimination:" << endl;
    print_grammar(result2);

    /* --- Example 3: Multiple recursive alternatives --- */
    cout << "\n----------------------------------------------------------------------" << endl;
    cout << "\n--- Example 3: Multiple Recursive Alternatives ---\n" << endl;
    Grammar grammar3 = parse_grammar(
        "A -> A b c | A d | e | f g\n"
    );
    cout << "Original Grammar:" << endl;
    print_grammar(grammar3);

    Grammar result3 = eliminate_left_recursion(grammar3);
    cout << "\nAfter Left Recursion Elimination:" << endl;
    print_grammar(result3);

    cout << "\n======================================================================" << endl;
    return 0;
}

File: exercises/python/04b_left_recursion/left_recursion.py
"""
================================================================================
Exercise 4b: Left Recursion Elimination
================================================================================

CONCEPT:
Left recursion in a context-free grammar occurs when a non-terminal A can
derive a sentential form starting with itself: A =>+ A alpha. This is
problematic for top-down (recursive descent / LL) parsers because they would
enter infinite loops trying to expand A. Left recursion comes in two forms:

  - Direct left recursion:   A -> A alpha | beta
  - Indirect left recursion: A -> B alpha, B -> A beta (A derives A through B)

ALGORITHM:
For direct left recursion, the transformation is straightforward:
  Original:    A -> A alpha1 | A alpha2 | ... | beta1 | beta2 | ...
  Transformed: A  -> beta1 A' | beta2 A' | ...
               A' -> alpha1 A' | alpha2 A' | ... | epsilon

For indirect left recursion, we use the standard ordering algorithm:
  1. Order all non-terminals as A1, A2, ..., An.
  2. For i = 1 to n:
       For j = 1 to i-1:
         Replace each production Ai -> Aj gamma with
         Ai -> delta1 gamma | delta2 gamma | ... where Aj -> delta1 | delta2 | ...
       Eliminate direct left recursion from Ai productions.

COMPLEXITY:
  Time:  O(n^2 * p) where n = number of non-terminals, p = max productions
  Space: O(n * p) for storing the grammar

EXAMPLE:
  Input:   E -> E+T | T,  T -> T*F | F,  F -> (E) | id
  Output:  E  -> T E'
           E' -> +T E' | epsilon
           T  -> F T'
           T' -> *F T' | epsilon
           F  -> (E) | id
================================================================================
"""

from collections import OrderedDict


def parse_grammar(grammar_text):
    """
    Parse a grammar string into an ordered dictionary.
    Each key is a non-terminal, each value is a list of productions (each production
    is a list of symbols).
    """
    grammar = OrderedDict()
    for line in grammar_text.strip().split("\n"):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        lhs, rhs = line.split("->")
        lhs = lhs.strip()
        alternatives = rhs.split("|")
        productions = []
        for alt in alternatives:
            symbols = alt.strip().split()
            productions.append(symbols)
        if lhs in grammar:
            grammar[lhs].extend(productions)
        else:
            grammar[lhs] = productions
    return grammar


def grammar_to_string(grammar):
    """Format a grammar dictionary as a readable string."""
    lines = []
    for nonterminal, productions in grammar.items():
        prod_strings = [" ".join(prod) for prod in productions]
        lines.append(f"  {nonterminal} -> {' | '.join(prod_strings)}")
    return "\n".join(lines)


def eliminate_direct_left_recursion(nonterminal, productions):
    """
    Eliminate direct left recursion for a single non-terminal.
    Given A -> A alpha1 | A alpha2 | ... | beta1 | beta2 | ...
    Produces:
      A  -> beta1 A' | beta2 A' | ...
      A' -> alpha1 A' | alpha2 A' | ... | epsilon
    Returns (new_productions_for_A, new_nonterminal_name, new_nonterminal_productions)
    or (original_productions, None, None) if no left recursion found.
    """
    recursive_prods = []     # productions starting with the nonterminal itself
    non_recursive_prods = [] # all other productions

    for prod in productions:
        if prod[0] == nonterminal:
            recursive_prods.append(prod[1:])  # alpha part (without leading A)
        else:
            non_recursive_prods.append(prod)

    if not recursive_prods:
        return productions, None, None

    # Create the new non-terminal name (A')
    new_nonterminal = nonterminal + "'"
    # Avoid collisions with existing names by adding more primes
    # (handled externally if needed)

    # A -> beta1 A' | beta2 A' | ...
    new_a_productions = []
    for prod in non_recursive_prods:
        if prod == ["epsilon"]:
            new_a_productions.append([new_nonterminal])
        else:
            new_a_productions.append(prod + [new_nonterminal])

    # A' -> alpha1 A' | alpha2 A' | ... | epsilon
    new_prime_productions = []
    for alpha in recursive_prods:
        if alpha:  # alpha is non-empty
            new_prime_productions.append(alpha + [new_nonterminal])
        else:
            new_prime_productions.append([new_nonterminal])
    new_prime_productions.append(["epsilon"])

    return new_a_productions, new_nonterminal, new_prime_productions


def substitute_productions(grammar, nonterminal_i, nonterminal_j):
    """
    For every production of Ai that starts with Aj, replace it by substituting
    all productions of Aj in place of Aj.
    Ai -> Aj gamma becomes Ai -> delta1 gamma | delta2 gamma | ...
    where Aj -> delta1 | delta2 | ...
    """
    new_productions = []
    for prod in grammar[nonterminal_i]:
        if prod[0] == nonterminal_j:
            # Substitute: replace Aj with each of its alternatives
            gamma = prod[1:]  # the part after Aj
            for delta in grammar[nonterminal_j]:
                if delta == ["epsilon"]:
                    if gamma:
                        new_productions.append(gamma)
                    else:
                        new_productions.append(["epsilon"])
                else:
                    new_productions.append(delta + gamma)
        else:
            new_productions.append(prod)
    grammar[nonterminal_i] = new_productions


def eliminate_left_recursion(grammar):
    """
    Eliminate all left recursion (both direct and indirect) using the
    standard ordering algorithm.
    """
    nonterminals = list(grammar.keys())
    new_grammar = OrderedDict(grammar)

    for i, nt_i in enumerate(nonterminals):
        # For each earlier non-terminal, substitute its productions
        for j in range(i):
            nt_j = nonterminals[j]
            substitute_productions(new_grammar, nt_i, nt_j)

        # Now eliminate any direct left recursion for nt_i
        new_prods, new_nt, new_nt_prods = eliminate_direct_left_recursion(
            nt_i, new_grammar[nt_i]
        )
        new_grammar[nt_i] = new_prods
        if new_nt is not None:
            # Insert the new non-terminal right after nt_i in the ordered dict
            items = list(new_grammar.items())
            idx = list(new_grammar.keys()).index(nt_i) + 1
            items.insert(idx, (new_nt, new_nt_prods))
            new_grammar = OrderedDict(items)

    return new_grammar


def main():
    print("=" * 70)
    print("  Exercise 4b: Left Recursion Elimination")
    print("=" * 70)

    # --- Example 1: Classic expression grammar with direct left recursion ---
    grammar_text_1 = """
        E -> E + T | T
        T -> T * F | F
        F -> ( E ) | id
    """
    print("\n--- Example 1: Direct Left Recursion ---")
    grammar_1 = parse_grammar(grammar_text_1)
    print("\nOriginal Grammar:")
    print(grammar_to_string(grammar_1))

    result_1 = eliminate_left_recursion(grammar_1)
    print("\nAfter Left Recursion Elimination:")
    print(grammar_to_string(result_1))

    # --- Example 2: Indirect left recursion ---
    grammar_text_2 = """
        S -> A a | b
        A -> S c | d
    """
    print("\n" + "-" * 70)
    print("\n--- Example 2: Indirect Left Recursion ---")
    grammar_2 = parse_grammar(grammar_text_2)
    print("\nOriginal Grammar:")
    print(grammar_to_string(grammar_2))

    result_2 = eliminate_left_recursion(grammar_2)
    print("\nAfter Left Recursion Elimination:")
    print(grammar_to_string(result_2))

    # --- Example 3: No left recursion (should remain unchanged) ---
    grammar_text_3 = """
        S -> a B | c
        B -> b S | d
    """
    print("\n" + "-" * 70)
    print("\n--- Example 3: No Left Recursion (unchanged) ---")
    grammar_3 = parse_grammar(grammar_text_3)
    print("\nOriginal Grammar:")
    print(grammar_to_string(grammar_3))

    result_3 = eliminate_left_recursion(grammar_3)
    print("\nAfter Left Recursion Elimination:")
    print(grammar_to_string(result_3))

    # --- Example 4: Multiple recursive alternatives ---
    grammar_text_4 = """
        A -> A b c | A d | e | f g
    """
    print("\n" + "-" * 70)
    print("\n--- Example 4: Multiple Recursive Alternatives ---")
    grammar_4 = parse_grammar(grammar_text_4)
    print("\nOriginal Grammar:")
    print(grammar_to_string(grammar_4))

    result_4 = eliminate_left_recursion(grammar_4)
    print("\nAfter Left Recursion Elimination:")
    print(grammar_to_string(result_4))

    print("\n" + "=" * 70)


if __name__ == "__main__":
    main()
